src/platform/packages/shared/kbn-grok-ui/models/grok_collection_and_pattern.ts

```diff
@@ -0,0 +1,393 @@
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+/* eslint-disable max-classes-per-file */
+
+import { toRegExp, toRegExpDetails } from 'oniguruma-to-es';
+import { monaco } from '@kbn/monaco';
+import { v4 as uuidv4 } from 'uuid';
+import { unflattenObject } from '@kbn/object-utils';
+import { euiPaletteColorBlindBehindText } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+
+// Grok patterns use this official naming: %{SYNTAX:SEMANTIC:TYPE}
+
+// Will match %{SYNTAX}, %{SYNTAX:SEMANTIC}, %{SYNTAX:SEMANTIC:TYPE}, and support special characters and dots.
+const SUBPATTERNS_REGEX =
+  /%\{[A-Z0-9_@#$%&*+=\-\.]+(?::[A-Za-z0-9_@#$%&*+=\-\.]+)?(?::[A-Za-z]+)?\}/g;
+
+// Matches "manual" semantic names in the expression, these are user defined capture groups, e.g. (?<field_name>the pattern here)
+const NESTED_FIELD_NAMES_REGEX =
+  /(\(\?<([A-Za-z0-9_@#$%&*+=\-\.]+)(?::([A-Za-z0-9_@#$%&*+=\-\.]+))?(?::([A-Za-z]+))?>)|\(\?:|\(\?>|\(\?!|\(\?<!|\(|\\\(|\\\)|\)|\[|\\\[|\\\]|\]/g;
+
+// The only supported semantic conversions are int and float. By default all semantics are saved as strings.
+// https://www.elastic.co/guide/en/logstash/current/plugins-filters-grok.html#_grok_basics
+const SUPPORTED_TYPE_CONVERSIONS = ['int', 'float'];
+
+// We supply this suffix to track which capture groups we have generated.
+const CAPTURE_GROUP_GENERATED_ID_SUFFIX = '_____GENERATED_CAPTURE_GROUP_____';
+
+const CUSTOM_NAMED_CAPTURE_PATTERN_PREFIX = i18n.translate(
+  'kbn.grokUi.customNamedCapturePatternPrefix',
+  { defaultMessage: 'Custom named capture ' }
+);
+
+export class GrokCollection {
+  private patterns: Map<string, GrokPattern> = new Map();
+  private patternKeys: string[] = [];
+  // NOTE: This doesn't subscribe to EUI_VIS_COLOR_STORE changes at the moment, whilst UI / UX is being finalised.
+  private colourPalette = euiPaletteColorBlindBehindText({ rotations: 3 });
+  private colourIndex = 0;
+
+  public getPattern(id: string) {
+    return this.patterns.get(id);
+  }
+
+  public addPattern(id: string, rawPattern: string) {
+    if (this.patterns.has(id)) {
+      // eslint-disable-next-line no-console
+      console.warn('Warning: pattern with ID: %s already exists', id);
+    } else {
+      const pattern = new GrokPattern(rawPattern, id, this);
+      this.patterns.set(id, pattern);
+
+      return pattern;
+    }
+  }
+
+  public resolvePatterns() {
+    this.patterns.forEach((pattern) => {
+      if (!pattern.isResolved()) {
+        pattern.resolvePattern();
+      }
+    });
+    this.patternKeys = Array.from(this.patterns.keys());
+  }
+
+  // Only relevant for Monaco users.
+  // Can be used with Monaco code editor to provide suggestions.
+  public getSuggestionProvider = () => {
+    const provider: monaco.languages.CompletionItemProvider = {
+      triggerCharacters: ['{'],
```

Comment:
Should `%{` be the trigger? `{` is just a normal character in grok, so it could trigger at the wrong time.

Reply:
It's kind of hard to find detailed information on [how they work](https://microsoft.github.io/monaco-editor/typedoc/interfaces/languages.CompletionItemProvider.html#triggerCharacters), but the behaviour is the same whether this is `{` or `%{`. I think we'd probably have to do a bunch of manual inspection via the text model if this becomes a problem (nothing is actually *selected* even if things displaying might not be necessary).